cloud/google/gceproviderconfig/v1alpha1/register.go

```diff
@@ -80,27 +83,23 @@ func NewCodec() (*GCEProviderConfigCodec, error) {
 	return &codec, nil
 }
 
-func (codec *GCEProviderConfigCodec) DecodeFromProviderConfig(providerConfig clusterv1.ProviderConfig) (*GCEProviderConfig, error) {
-	obj, gvk, err := codec.decoder.Decode(providerConfig.Value.Raw, nil, nil)
+func (codec *GCEProviderConfigCodec) DecodeFromProviderConfig(providerConfig clusterv1.ProviderConfig, out runtime.Object) (error) {
+	_, _, err := codec.decoder.Decode(providerConfig.Value.Raw, nil, out)
 	if err != nil {
-		return nil, fmt.Errorf("decoding failure: %v", err)
-	}
-	config, ok := obj.(*GCEProviderConfig)
-	if !ok {
-		return nil, fmt.Errorf("failure to cast to gce; type: %v", gvk)
+		return fmt.Errorf("decoding failure: %v", err)
 	}
-	return config, nil
+	return nil
 }
 
-func (codec *GCEProviderConfigCodec) EncodeToProviderConfig(gceProviderConfig *GCEProviderConfig) (*clusterv1.ProviderConfig, error) {
+func (codec *GCEProviderConfigCodec) EncodeToProviderConfig(in runtime.Object, out *clusterv1.ProviderConfig) (error) {
 	var buf bytes.Buffer
-	if err := codec.encoder.Encode(gceProviderConfig, &buf); err != nil {
-		return nil, fmt.Errorf("encoding failed: %v", err)
+	if err := codec.encoder.Encode(in, &buf); err != nil {
+		return fmt.Errorf("encoding failed: %v", err)
 	}
-	providerConfig := clusterv1.ProviderConfig{
+	*out = clusterv1.ProviderConfig{
```

Comment:
The motivation is to make this method work for both machines and clusters. Alternatively I could make 2 methods which return the correct type. @krousey suggested something like this in your [disk type PR](https://github.com/kubernetes-sigs/cluster-api/pull/77#discussion-diff-188379142R83), is this what you had in mind Kris?

Reply:
I understand what you are saying with regards to the Decode method and why it would require an generic out parameter, but why does that apply to the Encode method?

The return type before was "*clusterv1.ProviderConfig" which is now the type of the out parameter -- if all a caller does is declare a "var clusterv1.ProviderConfig" and then pass it in, i'm confused at how that changes the behavior in a way that would not be possible with an out parameter.

Reply:
Oh I see what you're saying. Yea, I agree, that makes more sense. I'll update this.